addon/model.js

```diff
@@ -102,41 +103,50 @@ const retrieveFromCurrentState = computed('_topModel.currentState', function(key
   return this._topModel._internalModel.currentState[key];
 }).readOnly();
 
-// global buffer for initial properties to work around
-//  a)  can't write to `this` before `super`
-//  b)  core_object writes properties before calling `init`; this means that no
-//      CP or setknownProperty can rely on any initialization
-let initProperites = Object.create(null);
+const INIT_KEYS = {
+  store: 1,
+  _internalModel: 1,
+  _topModel: 1,
+  _parentModel: 1,
+  isError: 1,
+};
 
-export default class MegamorphicModel extends EmberObject {
+export default class MegamorphicModel {
+  constructor(properties) {
+    this.init(properties);
+  }
   init(properties) {
-    // Drop Ember.Object subclassing instead
-    super.init(...arguments);
+    this._init = false;
+    this.store = properties.store;
     this._store = properties.store;
     this._internalModel = properties._internalModel;
     this._cache = Object.create(null);
     this._schema = get(properties.store, '_schemaManager');
 
-    this._topModel = this._topModel || this;
-    this._parentModel = this._parentModel || null;
-    this._errors = new DS.Errors();
-    this._init = true;
-
-    this._flushInitProperties();
-  }
-
-  _flushInitProperties() {
-    let propertiesToFlush = initProperites;
-    initProperites = Object.create(null);
+    this._topModel = properties._topModel || this;
+    this._parentModel = properties._parentModel || null;
+    this._errors = DS.Errors.create();
```

Comment:
can we lazily create this?

Reply:
Yes, and we should move to that for sure